[JSC] optimize `Promise.withResolvers` to avoid unnecessary object construction

dcrousso opened the pull request:
#### 0aebf750430c0f2bf1ad6b96a6d38c9640c63056
<pre>
[JSC] optimize `Promise.withResolvers` to avoid unnecessary object construction
<a href="https://bugs.webkit.org/show_bug.cgi?id=259175">https://bugs.webkit.org/show_bug.cgi?id=259175</a>

Reviewed by Yusuke Suzuki.

Rework `newPromiseCapability` (and `newPromiseCapabilitySlow`) to return an object containing `promise`/`resolve`/`reject` instead of `@promise`/`@resolve`/`@reject` so that `Promise.withResolvers` can directly use it instead of transforming it into another object.

As a result, update all the other callsites that use `@promise`/`@resolve`/`@reject` to use `promise`/`resolve`/`reject` as well.

* Source/JavaScriptCore/builtins/ModuleLoader.js:
(fulfillFetch):
* Source/JavaScriptCore/builtins/PromiseConstructor.js:
(promiseAllSlow):
(promiseNewOnRejected):
(allSettled):
(any):
(race):
(withResolvers):
* Source/JavaScriptCore/builtins/PromiseOperations.js:
(newPromiseCapabilitySlow):
(newPromiseCapability):
(promiseResolveSlow):
(promiseRejectSlow):
(createResolvingFunctions):
(createResolvingFunctionsWithoutPromise):
(promiseReactionJob):
(promiseResolveThenableJob):
(promiseResolveThenableJobWithDerivedPromise):
* Source/JavaScriptCore/builtins/PromisePrototype.js:
(then):
* Source/JavaScriptCore/runtime/JSPromise.cpp:
(JSC::JSPromise::convertCapabilityToDeferredData):
* Source/WebCore/Modules/streams/ReadableStreamBYOBReader.js:
(closed):
* Source/WebCore/Modules/streams/ReadableStreamDefaultReader.js:
(closed):
* Source/WebCore/Modules/streams/ReadableStreamInternals.js:
(readableStreamReaderGenericInitialize):
(readableStreamPipeTo):
(readableStreamPipeToWritableStream):
(pipeToErrorsMustBePropagatedForward):
(pipeToErrorsMustBePropagatedBackward):
(pipeToClosingMustBePropagatedForward):
(pipeToShutdownWithAction):
(pipeToShutdown):
(pipeToFinalize):
(readableStreamTee):
(pullAlgorithm):
(readableStreamTeePullFunction):
(readableStreamTeeBranch1CancelFunction):
(readableStreamTeeBranch2CancelFunction):
(readableStreamError):
(readableStreamClose):
(readableStreamReaderGenericRelease):
* Source/WebCore/Modules/streams/TransformStreamInternals.js:
(createInternalTransformStreamFromTransformer):
(createTransformStream):
(transformStreamSetBackpressure):
(transformStreamDefaultControllerPerformTransform):
(transformStreamDefaultSinkWriteAlgorithm):
(transformStreamDefaultSinkCloseAlgorithm):
(transformStreamDefaultSourcePullAlgorithm):
* Source/WebCore/Modules/streams/WritableStreamDefaultWriter.js:
(closed):
(ready):
* Source/WebCore/Modules/streams/WritableStreamInternals.js:
(setUpWritableStreamDefaultWriter):
(writableStreamAbort):
(writableStreamClose):
(writableStreamAddWriteRequest):
(writableStreamFinishErroring):
(writableStreamFinishInFlightWrite):
(writableStreamFinishInFlightWriteWithError):
(writableStreamRejectCloseAndClosedPromiseIfNeeded):
(writableStreamUpdateBackpressure):
(writableStreamDefaultWriterEnsureClosedPromiseRejected):
(writableStreamDefaultWriterEnsureReadyPromiseRejected):

Canonical link: <a href="https://commits.webkit.org/266036@main">https://commits.webkit.org/266036@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/54d119def17a339cd1f6f0461a4468534f84a0aa

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/12614 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/26/builds/12943 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/13258 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/14354 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/12082 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/12675 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/23/builds/15445 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/16/builds/12961 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/14779 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/12778 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/15/builds/13551 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/10676 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/14793 "Built successfully") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/39/builds/10828 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/11420 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/18509 "Passed tests") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/10723 "Built successfully and passed tests") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/11913 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/11585 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/14782 "Passed tests") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/11937 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/7/builds/12063 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/17/builds/9979 "Passed tests") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/12671 "Built successfully") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/30/builds/11309 "Built successfully") | | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/3317 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/3103 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/4/builds/15641 "Built successfully") | | [✅ 🛠 jsc-mips](https://ews-build.webkit.org/#/builders/24/builds/13027 "Built successfully") | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/31/builds/11919 "Built successfully") | | [✅ 🧪 jsc-mips-tests](https://ews-build.webkit.org/#/builders/3/builds/3105 "Passed tests") | 
<!--EWS-Status-Bubble-End-->

Constellation commented on Source/JavaScriptCore/builtins/PromiseOperations.js:
```diff
@@ -91,13 +91,13 @@ function newPromiseCapability(constructor)
         var promise = @newPromise();
         var capturedPromise = promise;
-        function @resolve(resolution) {
+        function resolve(resolution) {
```
We need to keep this `@resolve` to make this unnamed function (we have special path for private name functions).

Constellation commented on Source/JavaScriptCore/builtins/PromiseOperations.js:
```diff
@@ -91,13 +91,13 @@ function newPromiseCapability(constructor)
             return @resolvePromiseWithFirstResolvingFunctionCallCheck(capturedPromise, resolution);
         }
-        function @reject(reason) {
+        function reject(reason) {
```
We need to keep this `@resolve` to make this unnamed function (we have special path for private name functions).

Constellation reviewed (approved):
r=me with some changes.

dcrousso committed:
[JSC] optimize `Promise.withResolvers` to avoid unnecessary object construction
https://bugs.webkit.org/show_bug.cgi?id=259175

Reviewed by Yusuke Suzuki.

Rework `newPromiseCapability` (and `newPromiseCapabilitySlow`) to return an object containing `promise`/`resolve`/`reject` instead of `@promise`/`@resolve`/`@reject` so that `Promise.withResolvers` can directly use it instead of transforming it into another object.

As a result, update all the other callsites that use `@promise`/`@resolve`/`@reject` to use `promise`/`resolve`/`reject` as well.

* Source/JavaScriptCore/builtins/ModuleLoader.js:
(fulfillFetch):
* Source/JavaScriptCore/builtins/PromiseConstructor.js:
(promiseAllSlow):
(promiseNewOnRejected):
(allSettled):
(any):
(race):
(withResolvers):
* Source/JavaScriptCore/builtins/PromiseOperations.js:
(newPromiseCapabilitySlow):
(newPromiseCapability):
(promiseResolveSlow):
(promiseRejectSlow):
(createResolvingFunctions):
(createResolvingFunctionsWithoutPromise):
(promiseReactionJob):
(promiseResolveThenableJob):
(promiseResolveThenableJobWithDerivedPromise):
* Source/JavaScriptCore/builtins/PromisePrototype.js:
(then):
* Source/JavaScriptCore/runtime/JSPromise.cpp:
(JSC::JSPromise::convertCapabilityToDeferredData):
* Source/WebCore/Modules/streams/ReadableStreamBYOBReader.js:
(closed):
* Source/WebCore/Modules/streams/ReadableStreamDefaultReader.js:
(closed):
* Source/WebCore/Modules/streams/ReadableStreamInternals.js:
(readableStreamReaderGenericInitialize):
(readableStreamPipeTo):
(readableStreamPipeToWritableStream):
(pipeToErrorsMustBePropagatedForward):
(pipeToErrorsMustBePropagatedBackward):
(pipeToClosingMustBePropagatedForward):
(pipeToShutdownWithAction):
(pipeToShutdown):
(pipeToFinalize):
(readableStreamTee):
(pullAlgorithm):
(readableStreamTeePullFunction):
(readableStreamTeeBranch1CancelFunction):
(readableStreamTeeBranch2CancelFunction):
(readableStreamError):
(readableStreamClose):
(readableStreamReaderGenericRelease):
* Source/WebCore/Modules/streams/TransformStreamInternals.js:
(createInternalTransformStreamFromTransformer):
(createTransformStream):
(transformStreamSetBackpressure):
(transformStreamDefaultControllerPerformTransform):
(transformStreamDefaultSinkWriteAlgorithm):
(transformStreamDefaultSinkCloseAlgorithm):
(transformStreamDefaultSourcePullAlgorithm):
* Source/WebCore/Modules/streams/WritableStreamDefaultWriter.js:
(closed):
(ready):
* Source/WebCore/Modules/streams/WritableStreamInternals.js:
(setUpWritableStreamDefaultWriter):
(writableStreamAbort):
(writableStreamClose):
(writableStreamAddWriteRequest):
(writableStreamFinishErroring):
(writableStreamFinishInFlightWrite):
(writableStreamFinishInFlightWriteWithError):
(writableStreamRejectCloseAndClosedPromiseIfNeeded):
(writableStreamUpdateBackpressure):
(writableStreamDefaultWriterEnsureClosedPromiseRejected):
(writableStreamDefaultWriterEnsureReadyPromiseRejected):

Canonical link: https://commits.webkit.org/266036@main
